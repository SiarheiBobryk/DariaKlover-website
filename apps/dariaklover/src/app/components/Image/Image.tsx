import { styled } from '@mui/material/styles';
import Typography from '@mui/material/Typography';
import * as React from 'react';

const { NX_IMAGES_PATH = '' } = process.env;

export const Img = styled('img')(function styleImg({ theme }) {
  return {
    maxWidth: '100%',
    borderRadius: theme.typography.pxToRem(8),
  };
});

export const Figure = styled('figure')(function styleFigure({ theme }) {
  return {
    marginTop: 0,
    marginRight: 0,
    marginBottom: theme.typography.pxToRem(8),
    marginLeft: 0,
  };
});

export interface ImageProps {
  fileName: string;
  alt: string;
  caption: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Image = React.forwardRef(function Image(props: ImageProps, ref: React.Ref<any>) {
  const { fileName, alt, caption, ...other } = props;

  return (
    <Figure ref={ref} {...other}>
      <picture>
        <source type="image/webp" srcSet={`${NX_IMAGES_PATH}/${fileName}.webp`} />
        <Img loading="lazy" src={`${NX_IMAGES_PATH}/${fileName}.png`} alt={alt} />
      </picture>
      <figcaption>
        <Typography component="p" variant="caption" sx={{ textAlign: 'center' }}>
          {caption}
        </Typography>
      </figcaption>
    </Figure>
  );
});

export default Image;
