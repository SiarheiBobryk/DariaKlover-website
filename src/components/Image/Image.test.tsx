import * as React from 'react';

import { axe, toHaveNoViolations } from 'jest-axe';
import { render, RenderResult, screen } from '@testing-library/react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';

import Image from './Image';

expect.extend(toHaveNoViolations);

describe('<Image />', () => {
  it('should match a11y standards without props', async () => {
    const { container }: RenderResult = render(
      <Image fileName="how-to-book-1" alt="Установка часового пояса" caption="Установка вашего часового пояса" />,
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should forward ref', () => {
    const ref = React.createRef();
    render(
      <Image
        ref={ref}
        data-testid="Image"
        fileName="how-to-book-1"
        alt="Установка часового пояса"
        caption="Установка вашего часового пояса"
      />,
    );
    expect(screen.getByTestId('Image')).toBe(ref.current);
  });

  it('should render correctly', () => {
    const image: ReactTestRendererJSON | ReactTestRendererJSON[] | null = renderer
      .create(
        <Image fileName="how-to-book-1" alt="Установка часового пояса" caption="Установка вашего часового пояса" />,
      )
      .toJSON();
    expect(image).toMatchSnapshot();
  });
});
