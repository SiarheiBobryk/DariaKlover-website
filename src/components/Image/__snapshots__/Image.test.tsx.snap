// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> should render correctly 1`] = `
<figure
  className="css-b9qyun"
>
  <picture>
    <source
      srcSet="responsive_images/how-to-book-1.webp"
      type="image/webp"
    />
    <img
      alt="Установка часового пояса"
      className="css-gkqd5w"
      loading="lazy"
      src="responsive_images/how-to-book-1.png"
    />
  </picture>
  <figcaption>
    <p
      className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
    >
      Установка вашего часового пояса
    </p>
  </figcaption>
</figure>
`;
